### PR TITLE
add  flecs/v2.3.2

### LINKED ISSUE #4541: [request] flecs/v2.3.1
### Package Details
  * Package Name/Version: **flecs/v2.3.1**
  * Website: **https://github.com/SanderMertens/flecs**
  * Source code: **https://github.com/SanderMertens/flecs**
  * Platform/Compiler:
     * Windows: msvc
     * Ubuntu: gcc (7, 8, 9, 10), clang (8, 9)
     * MacOS: gcc 10, clang 9
  * Conan recipe repository: https://github.com/rawbby/conan-flecs
    (As I am new to the Conan Community I dont know if the package is perfectly polished yet)

### Description Of The Library / Tool
Flecs is a fast and lightweight Entity Component System with a focus on high performance game development.

### PR DESCRIPTION
**Pull request by rawbby**
Specify library name and version: **flecs/v2.3.2** (not **flecs/v2.3.1** anymore)

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

closes #4541

### CODE REVIEW

**Comment by SpaceIm** [comment]
Please consider https://github.com/rawbby/conan-center-index/pull/1

**Comment by rawbby** [comment]
Hey @SpaceIm,

I've merged your changes. Thanks a lot for your work!
I ask the author of flecs to write me the most important topics.

[question]
It would be nice if the flecs_STATIC compile definition would be set by the conan target for cmake.
I dont know if this is possible. Can some tell me if this is possible?

**Comment by SpaceIm** [comment]
> [question]
> It would be nice if the flecs_STATIC compile definition would be set by the conan target for cmake.
> I dont know if this is possible. Can some tell me if this is possible?

I did it in `package_info()`:

```python
        if not self.options.shared:
            self.cpp_info.defines.append("flecs_STATIC")
```

**Comment by rawbby** [comment]
@SpaceIm can you tell why strcat_s is unresolved for the above compilers?

**Comment by rawbby** [comment]
Is there a way to tell conan, not to use gcc<5?

**Comment by SpaceIm** [comment]
yes, but I would define `FLECS_NO_CPP` in test_package and make a pure C example, because this error comes from the C++ API which can be disabled at consume time.

**Comment by SpaceIm** [comment]
I found the issues, I'll submit a PR to your branch.

**Comment by SpaceIm** [comment]
Please merge https://github.com/rawbby/conan-center-index/pull/2, it should fix everything

**Comment by rawbby** [comment]
@SpaceIm how do you determine the right cmake version?
As far as I know CMake changes behavior in some versions. Therefor I tend to use up to date versions of CMake to provide up to date behavior. Am I wrong here? (why?)

**Comment by SpaceIm** [comment]
You are not wrong, but for open source libraries too high CMake versions might be an issue for consumers. Yes CMake can change behavior of several functions or features. It it happens, you can set a policy to revert to the old behavior so that know that it will provide a consistent bheavior accros versions.
In what I saw in upstream CMakeLists, there is nothing very new, so I'm pretty sure it should run the same way with CMake 3.1 and 3.19 (but it's always hard to know everything about CMake).

**Review by SpaceIm — APPROVED** [review]
I've submitted https://github.com/SanderMertens/flecs/pull/320 upstream

**Comment by SpaceIm** [comment]
Well it has been merged quickly, maybe we can just add the same modification as a patch, and simplify conanfile.py to rely on new options and instal target.

**Comment by prince-chrismc** [comment]
> Well it has been merged quickly

Forgot to look at the tag version when I  started my review.... no wonder the CMake looks so good.

 Great work! 🏆

File: recipes/flecs/all/conandata.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+sources:
+  "2.3.1":
+    url: "https://github.com/SanderMertens/flecs/archive/v2.3.1.tar.gz"
+    sha256: "b2cd69b3282167295d509deb762e580ad76040ed96ceb4492657bc1e444b8c89"
+patches:
+  "2.3.1":
+    - patch_file: "patches/less-warnings.patch"
```
Comment (prince-chrismc) [inline recipes/flecs/all/conandata.yml:7]: looking at the patch and the conversation it does not seem to be required..

I think patch could be removed

**Review by prince-chrismc — COMMENTED** [review]
Would it be worth moving this to a specific subfolder?

The next release would be a fairly different recipe

File: recipes/flecs/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        cmake.build(target="flecs" if self.options.shared else "flecs_static")
+
+    def package(self):
+        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
+        self.copy("*", dst="include", src=os.path.join(self._source_subfolder, "include"))
+        self.copy("*.lib", dst="lib", src="lib")
+        self.copy("*.dll", dst="bin", src="bin")
+        self.copy("*.a", dst="lib", src="lib")
+        self.copy("*.so*", dst="lib", src="lib", symlinks=True)
+        self.copy("*.dylib", dst="lib", src="lib", symlinks=True)
```
Comment (uilianries) [inline recipes/flecs/all/conanfile.py:54]:
```suggestion
    def _configure_cmake(self):
        if self._cmake:
            return self._cmake
        self._cmake = CMake(self)
        self._cmake.configure()        
        return self._cmake

    def build(self):
        for patch in self.conan_data.get("patches", {}).get(self.version, []):
            tools.patch(**patch)
        cmake = self._configure_cmake()
        cmake.build(target="flecs" if self.options.shared else "flecs_static")

    def package(self):
        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
        cmake = self._configure_cmake()
        cmake.install()
```

File: recipes/flecs/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+              "cpp11", "ecs-framework")
+    homepage = "https://github.com/SanderMertens/flecs"
+    url = "https://github.com/conan-io/conan-center-index"
+
+    settings = "os", "arch", "compiler", "build_type"
+    options = {"shared": [True, False], "fPIC": [True, False]}
+    default_options = {"shared": False, "fPIC": True}
+
+    exports_sources = ["CMakeLists.txt", "patches/**"]
+    generators = "cmake"
```
Comment (uilianries) [inline recipes/flecs/all/conanfile.py:20]:
```suggestion
    generators = "cmake"
    _cmake = None
```

File: recipes/flecs/all/patches/less-warnings.patch [hunk 4]
```diff
@@ -0,0 +1,32 @@
+--- a/cmake/target_default_compile_warnings.cmake
```
Comment (uilianries) [inline recipes/flecs/all/patches/less-warnings.patch:1]: It should be removed, it doesn't patch the building.

**Review by uilianries — APPROVED** [review]

**Review by prince-chrismc — DISMISSED** [review]

**Comment by rawbby** [comment]
The commit name was misleading. It is a remove invalid compiler warnings patch...

File: recipes/flecs/all/patches/less-warnings.patch (hunk 4 above)
Comment (rawbby) [reply recipes/flecs/all/patches/less-warnings.patch]: > 
> 
> The commit/patch name is misleading. It removes invalid compiler warnings...

File: recipes/flecs/all/conandata.yml (hunk 1 above)
Comment (rawbby) [reply recipes/flecs/all/conandata.yml]: > 
> 
> The commit/patch name is misleading. It removes invalid compiler warnings...

File: recipes/flecs/all/conanfile.py (hunk 2 above)
Comment (rawbby) [reply recipes/flecs/all/conanfile.py]: I think the default install is not working with the current cmake version.
We may have the revert these changes.

**Comment by rawbby** [comment]
I asked the flecs author for a new TAG and he wants to create one today. (with the new cmake patches)
Maybe we want to not support v2.3.1 at all and start support for v2.3.2 onward.

**Comment by uilianries** [comment]
> I asked the flecs author for a new TAG and he wants to create one today. (with the new cmake patches)
> Maybe we want to not support v2.3.1 at all and start support for v2.3.2 onward.

It sounds great, if we can move to a new tag with better support.

File: recipes/flecs/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+              "game-development", "ecs", "entity-component-system",
+              "cpp11", "ecs-framework")
+    homepage = "https://github.com/SanderMertens/flecs"
+    url = "https://github.com/conan-io/conan-center-index"
+
+    settings = "os", "arch", "compiler", "build_type"
+    options = {"shared": [True, False], "fPIC": [True, False]}
+    default_options = {"shared": False, "fPIC": True}
+
+    exports_sources = ["CMakeLists.txt", "patches/**"]
```
Comment (SpaceIm) [inline recipes/flecs/all/conanfile.py:19]:
```suggestion
    exports_sources = "CMakeLists.txt"
```

File: recipes/flecs/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        del self.settings.compiler.cppstd
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        os.rename(self.name + "-" + self.version, self._source_subfolder)
+
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+        self._cmake = CMake(self)
```
Comment (SpaceIm) [inline recipes/flecs/all/conanfile.py:44]:
```suggestion
        self._cmake = CMake(self)
        self._cmake.definitions["FLECS_STATIC_LIBS"] = not self.options.shared
        self._cmake.definitions["FLECS_PIC"] = self.options.get_safe("fPIC", True)
        self._cmake.definitions["FLECS_SHARED_LIBS"] = self.options.shared
        self._cmake.definitions["FLECS_DEVELOPER_WARNINGS"] = False
```

File: recipes/flecs/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        self._cmake.configure()
+        return self._cmake
+
+    def build(self):
+        for patch in self.conan_data.get("patches", {}).get(self.version, []):
+            tools.patch(**patch)
```
Comment (SpaceIm) [inline recipes/flecs/all/conanfile.py:50]:
```suggestion
```

File: recipes/flecs/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        cmake = self._configure_cmake()
+        cmake.build(target="flecs" if self.options.shared else "flecs_static")
```
Comment (SpaceIm) [inline recipes/flecs/all/conanfile.py:52]:
```suggestion
        cmake.build()
```

File: recipes/flecs/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+    def package(self):
+        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
+        cmake = self._configure_cmake()
+        cmake.install()
```
Comment (SpaceIm) [inline recipes/flecs/all/conanfile.py:57]:
```suggestion
        cmake.install()
        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
```

File: recipes/flecs/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+    def package_info(self):
+        self.cpp_info.libs = ["flecs"] if self.options.shared else ["flecs_static"]
+        if not self.options.shared:
+            self.cpp_info.defines.append("flecs_STATIC")
```
Comment (SpaceIm) [inline recipes/flecs/all/conanfile.py:62]:
```suggestion
        self.cpp_info.names["cmake_find_package"] = "flecs"
        self.cpp_info.names["cmake_find_package_multi"] = "flecs"
        target_and_lib_name = "flecs" if self.options.shared else "flecs_static"
        self.cpp_info.components["_flecs"].names["cmake_find_package"] = target_and_lib_name
        self.cpp_info.components["_flecs"].names["cmake_find_package_multi"] = target_and_lib_name
        self.cpp_info.components["_flecs"].libs = [target_and_lib_name]
        if not self.options.shared:
            self.cpp_info.components["_flecs"].defines.append("flecs_STATIC")
```

File: recipes/flecs/all/test_package/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
+from conans import ConanFile, CMake, tools
+import os
+
+
+class TestPackageConan(ConanFile):
+    settings = "os", "arch", "compiler", "build_type"
+    generators = "cmake"
```
Comment (SpaceIm) [inline recipes/flecs/all/test_package/conanfile.py:7]:
```suggestion
    generators = "cmake", "cmake_find_package_multi"
```

File: recipes/flecs/all/test_package/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package C)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(${PROJECT_NAME} test_package.c)
+target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
```
Comment (SpaceIm) [inline recipes/flecs/all/test_package/CMakeLists.txt:8]:
```suggestion
conan_basic_setup(TARGETS)

find_package(flecs REQUIRED CONFIG)

add_executable(${PROJECT_NAME} test_package.c)
if(FLECS_SHARED)
  target_link_libraries(${PROJECT_NAME} flecs::flecs)
else()
  target_link_libraries(${PROJECT_NAME} flecs::flecs_static)
endif()
```

**Review by SpaceIm — CHANGES_REQUESTED** [review]
To properly handle new upstream CMakeLists

File: recipes/flecs/all/test_package/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,17 @@
+from conans import ConanFile, CMake, tools
+import os
+
+
+class TestPackageConan(ConanFile):
+    settings = "os", "arch", "compiler", "build_type"
+    generators = "cmake", "cmake_find_package_multi"
+
+    def build(self):
+        cmake = CMake(self)
```
Comment (SpaceIm) [inline recipes/flecs/all/test_package/conanfile.py:10]:
```suggestion
        cmake = CMake(self)
        cmake.definitions["FLECS_SHARED"] = self.options["flecs"].shared
```

Forgot that it needs also this in test_package https://github.com/SpaceIm/conan-flecs/blob/f633f0c705141e8eea726db9ca713bc780108f95/test_package/conanfile.py#L11

**Review by SpaceIm — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Review by intelligide — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

**Pull request merged**